lib/group-update.js

```diff
@@ -0,0 +1,335 @@
+
+                const jobOptions = {
+                    course_id: course_id,
+                    course_instance_id: course_instance_id,
+                    assessment_id: assessment_id,
+                    user_id: user_id,
+                    authn_user_id: authn_user_id,
+                    type: 'auto_generate_groups',
+                    description: 'Auto generate group settings for ' + assessment_label,
+                    job_sequence_id: job_sequence_id,
+                    last_in_sequence: true,
+                };
+                serverJobs.createJob(jobOptions, function(err, job) {
+                    if (err) {
+                        logger.error('Error in createJob()', err);
+                        serverJobs.failJobSequence(job_sequence_id);
+                        return;
+                    }
+                    (async () => {
+                        try{
+                            job.verbose('Auto generate group settings for ' + assessment_label);
+                            job.verbose(`----------------------------------------`);
+                            job.verbose(`Fetching the enrollment lists...`);
+                            const aid = {assessment_id};
+                            var students = [];
+                            const resultlist = await sqldb.queryAsync(sql.select_enrollments, aid);
+                            resultlist.rows.forEach(element => {
+                                students.push(element.user_list);
+                            });
+                            module.exports._shuffle(students);
+                            var num_students = resultlist.rowCount;
+                            var not_assigned = [];
+                            const resultlist2 = await sqldb.queryAsync(sql.select_not_assigned, aid);
+                            resultlist2.rows.forEach(element => {
+                                not_assigned.push(element.user_list);
+                            });
+                            module.exports._shuffle(not_assigned);
+                            var num_not_assigned = resultlist2.rowCount;
+                            job.verbose(`There are ` + num_students + ' students enrolled in ' + assessment_label);
+                            job.verbose(num_not_assigned + ' of them have not been in a group');
+                            job.verbose(`----------------------------------------`);
+                            job.verbose(`Processing creating groups - max of ` + maxnum + ' and min of ' + minnum);
+                            let num_group = Math.ceil(num_students / maxnum);
+                            let successCount = 0;
+                            let errorCount = 0;
+                            let output = null;
+                            let outputCount = 0;
+                            let outputThreshold = 100;
+                            for(let i = 0; i < num_group; i++){
```

Comment:
Code style: Please use spaces like `for () {` and `if () {`. This applies everywhere in this PR, not just in this file.